File: v1/scala/rail-fence-cipher/src/main/scala/RailFenceCipher.scala
object RailFenceCipher {
  def track(rails: Int, index: Int): Int = {
    val size = rails * 2 - 2
    def correct(i: Int): Boolean = i % size == 0

    if (correct(index)) 0
    else if (correct(index - rails + 1)) rails - 1
    else (1 to rails - 2).find(i => correct(index - i) || correct(index - size + i)).get
  }

  def encode(plainText: String, rails: Int): String = {
    val lettersByRail = (0 until plainText.length)
      .map(i => (track(rails, i), plainText.charAt(i)))
      .groupBy(_._1)

    lettersByRail
      .keys
      .toList
      .sorted
      .map(i => lettersByRail(i).map(a => a._2).mkString)
      .mkString
  }

  def decode(encodedText: String, rails: Int): String = {
    val railIndices = (0 until encodedText.length)
      .groupBy(i => track(rails, i))

    val letterIndices = railIndices
      .keys
      .toList
      .sorted
      .flatMap(railIndices(_))

    encodedText
      .zip(letterIndices)
      .sortBy(_._2)
      .map(_._1)
      .mkString
  }
}

File: haskell/rail-fence-cipher/src/RailFenceCipher.hs
module RailFenceCipher (encode, decode) where

import Data.List (find, filter, zip, sortOn)
import Data.List.Index (imap)
import Data.Maybe (fromJust)

track :: Int -> Int -> Int
track rails index = 
    if correct index then 0
    else if correct (index - rails + 1) then rails - 1
    else fromJust $ find (\i -> correct (index - i) || correct (index - size + i)) [1 .. (rails - 2)]
    where
        size = rails * 2 - 2
        correct i = i `mod` size == 0

encode :: Int -> String -> String
encode rails message = concatMap charsForRails [0 .. rails - 1]
    where
        charsWithTrack = imap (\index character -> (track rails index, character)) message
        charsForRails rail = map snd $ filter (\x -> fst x == rail) charsWithTrack

decode :: Int -> String -> String
decode rails message = map fst $ sortOn snd charsWithRail
    where
        railsForChars = map fst $ sortOn snd $ map (\i -> (i, track rails i)) [0 .. length message - 1]
        charsWithRail = zip message railsForChars